nasl/nasl_ssh.c

```diff
@@ -1743,6 +1743,50 @@ nasl_ssh_shell_open (lex_ctxt *lexic)
   return retc;
 }
 
+/**
+ * @brief read from an ssh channel until timeouts or there is no bytes left to
+ * read.
+ *
+ * @param[in]   channel     SSH Channel.
+ * @param[out]  response    Buffer to store response in.
+ * @param[in]   timeout     Timeout in milliseconds.
+ *
+ * @return 0 if success, -1 if error.
+ */
+static int
+read_ssh_blocking (ssh_channel channel, GString *response, int timeout)
+{
+  int rc;
+  char buffer[4096];
+
+  while (rc > 0 || rc == SSH_AGAIN)
+    {
+      if ((rc = ssh_channel_read_timeout (channel, buffer, sizeof (buffer), 1,
+                                          timeout))
+          > 0)
+        g_string_append_len (response, buffer, rc);
+      else if (rc == SSH_ERROR)
+        goto exec_err;
+      else if (rc == SSH_AGAIN)
+        continue;
+    }
+  rc = 1;
+  while (rc > 0 || rc == SSH_AGAIN)
+    {
+      if ((rc = ssh_channel_read_timeout (channel, buffer, sizeof (buffer), 0,
+                                          timeout))
+          > 0)
+        g_string_append_len (response, buffer, rc);
+
+      else if (rc == SSH_ERROR)
+        goto exec_err;
+    }
```

Comment:
What is the purpose of the second loop? Maybe it is not needed with the changes above?

Reply:
ssh_channel_read_timeout is called twice with the 3rd parameter `0` and `1` . These are stdout and stderr respectively.